extensions/gcp/controllerservices/GCPCredentialsControllerService.h

```diff
@@ -30,22 +30,44 @@
 #include "google/cloud/storage/oauth2/credentials.h"
 
 namespace org::apache::nifi::minifi::extensions::gcp {
+enum class CredentialsLocation {
+  USE_DEFAULT_CREDENTIALS,
+  USE_COMPUTE_ENGINE_CREDENTIALS,
+  USE_JSON_FILE,
+  USE_JSON_CONTENTS,
+  USE_ANONYMOUS_CREDENTIALS
+};
+}  // namespace org::apache::nifi::minifi::extensions::gcp
 
-SMART_ENUM(CredentialsLocation,
-    (USE_DEFAULT_CREDENTIALS, "Google Application Default Credentials"),
-    (USE_COMPUTE_ENGINE_CREDENTIALS, "Use Compute Engine Credentials"),
-    (USE_JSON_FILE, "Service Account JSON File"),
-    (USE_JSON_CONTENTS, "Service Account JSON"),
-    (USE_ANONYMOUS_CREDENTIALS, "Use Anonymous credentials"));
+namespace magic_enum::customize {
+template <>
+constexpr customize_t enum_name<org::apache::nifi::minifi::extensions::gcp::CredentialsLocation>(org::apache::nifi::minifi::extensions::gcp::CredentialsLocation value) noexcept {
+  switch (value) {
+    case org::apache::nifi::minifi::extensions::gcp::CredentialsLocation::USE_DEFAULT_CREDENTIALS:
+      return "Google Application Default Credentials";
+    case org::apache::nifi::minifi::extensions::gcp::CredentialsLocation::USE_COMPUTE_ENGINE_CREDENTIALS:
+      return "Use Compute Engine Credentials";
+    case org::apache::nifi::minifi::extensions::gcp::CredentialsLocation::USE_JSON_FILE:
+      return "Service Account JSON File";
+    case org::apache::nifi::minifi::extensions::gcp::CredentialsLocation::USE_JSON_CONTENTS:
+      return "Service Account JSON";
+    case org::apache::nifi::minifi::extensions::gcp::CredentialsLocation::USE_ANONYMOUS_CREDENTIALS:
+      return "Use Anonymous credentials";
+  }
+  return default_tag;
```

Comment:
In the examples provided by the `magic_enum` library the `default_tag` was used in these specialization, that's why I used the same, but I think you are right the default value could be misleading. My problem with the `invalid_tag` that it has a template specialization which returns an empty representation which has its `data()` member return a `nullptr` which could cause undefined behavior when used as a logger parameter. It may be better to return a unique custom value like `"INVALID_ENUM_NAME"` in this case, what do you think?

Reply:
I agree it may be better to use `std::string{sv}` instead of the `.data()` and use the `invalid_data` in custom enum name implementations. I updated it in 142a2b3b0b9599c6650844c73788cb124a15c291